pkg/river/internal/value/number_value.go

```diff
@@ -0,0 +1,120 @@
+package value
+
+import (
+	"math"
+	"reflect"
+	"strconv"
+)
+
+var (
+	nativeIntBits  = reflect.TypeOf(int(0)).Bits()
+	nativeUintBits = reflect.TypeOf(uint(0)).Bits()
+)
+
+// numberKind categorizes a type of Go number.
+type numberKind uint8
+
+const (
+	numberKindInt numberKind = iota
+	numberKindUint
+	numberKindFloat
+)
+
+func makeNumberKind(k reflect.Kind) numberKind {
+	switch k {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return numberKindInt
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return numberKindUint
+	case reflect.Float32, reflect.Float64:
+		return numberKindFloat
+	default:
+		panic("river/value: makeNumberKind called with unsupported Kind value")
+	}
+}
+
+// numberValue is a generic representation of Go numbers. It is intended to be
+// created on the fly for numerical operations when the real number type is not
+// known.
+type numberValue struct {
+	// Value holds the raw data for the number. Note that for numberKindFloat,
+	// value is the raw bits of the float64 and must be converted back to a
+	// float64 before it can be used.
+	value uint64
+
+	bits uint8      // 8, 16, 32, 64, used for overflow checking
+	k    numberKind // int, uint, float
+}
+
+func newNumberValue(v reflect.Value) numberValue {
+	var (
+		val  uint64
+		bits int
+		nk   numberKind
+	)
+
+	switch v.Kind() {
+	case reflect.Int:
+		val, bits, nk = uint64(v.Int()), nativeIntBits, numberKindInt
+	case reflect.Int8:
+		val, bits, nk = uint64(v.Int()), 8, numberKindInt
+	case reflect.Int16:
+		val, bits, nk = uint64(v.Int()), 16, numberKindInt
+	case reflect.Int32:
+		val, bits, nk = uint64(v.Int()), 32, numberKindInt
+	case reflect.Int64:
+		val, bits, nk = uint64(v.Int()), 64, numberKindInt
+	case reflect.Uint:
+		val, bits, nk = v.Uint(), nativeUintBits, numberKindUint
+	case reflect.Uint8:
+		val, bits, nk = v.Uint(), 8, numberKindUint
+	case reflect.Uint16:
+		val, bits, nk = v.Uint(), 16, numberKindUint
+	case reflect.Uint32:
+		val, bits, nk = v.Uint(), 32, numberKindUint
+	case reflect.Uint64:
+		val, bits, nk = v.Uint(), 64, numberKindUint
+	case reflect.Float32:
+		val, bits, nk = math.Float64bits(v.Float()), 32, numberKindFloat
+	case reflect.Float64:
+		val, bits, nk = math.Float64bits(v.Float()), 64, numberKindFloat
+	default:
+		panic("river/value: unrecognized Go number type " + v.Kind().String())
+	}
+
+	return numberValue{val, uint8(bits), nk}
+}
+
+func (nv numberValue) Int() int64 {
+	if nv.k == numberKindFloat {
+		return int64(math.Float64frombits(nv.value))
```

Comment:
I suppose this truncates the floating-point part. Maybe we want to raise an error here instead?

Reply:
It would implicitly truncate, yeah. If we make this return an error, then we'd have to give the user a function to manually convert from a float-type-number to an int-type-number, which may be tedious (or confusing since there's currently only one River number type which encompasses all the Go number types). 

I'd like to leave it as-is for now and revisit later if it becomes a source of confusion.

Reply:
Yeah, makes sense. Thanks!